Major gates.rst revision (#393)
openqasm/openqasm: source/language/gates.rst

Applying gates
--------------

Every gate applies to a fixed number of qubits.
Assuming ``g1``, ``g2``, ... are gates defined on 1, 2, ... qubits then they can be applied as

.. code-block::

   g1 q1;
   g2 q1, q2;
   g3 q1, q2, q3;

Broadcasting
~~~~~~~~~~~~

If any arguments of a gate are quantum registers instead of qubits, then all such registers must be **of the same length** and
this syntax is a convenient shorthand for broadcasting over the qubits of the register. For example, the circuit

.. code-block::

   qubit qr0[1];
   qubit qr1[3];
   qubit qr2[2];
   qubit qr3[3];

   g4 qr0[0], qr1, qr2[0], qr3; // ok
   g4 qr0[0], qr2, qr1[0], qr3; // error! qr2 and qr3 differ in size

has a second-to-last line that is equivalent to

.. code-block:: text

   g4 qr0[0], qr1[0], qr2[0], qr3[0];
   g4 qr0[0], qr1[1], qr2[0], qr3[1];
   g4 qr0[0], qr1[2], qr2[0], qr3[2];

In case that the expanded set of gates is not mutually-commuting, then its important to note that the broadcasting is defined
in increasing qubit index order ``qr[0]``, ``qr[1]``, ....

//FIXME: Is that what we really want? Maybe we could use attributes to mark which arguments are safe (commuting) to broadcast over

> and error in the case of non-commuting broadcasting?

Reviewer: I think this is what we agreed to, at least.

Parameterized gates
~~~~~~~~~~~~~~~~~~~

As well as gates that each represent a fixed unitary, OpenQASM also supports gates that represent *families* of unitaries, parameterized

> by angle variables.  Todistinguish the (optional) angle parameters from the (required) quantum arguments, if any angle parameters are

Reviewer 2: minor typo
```suggestion
by angle variables. To distinguish the (optional) angle parameters from the (required) quantum arguments, if any angle parameters are
```

   The definitions of these gates is part of the language specification.

The next subsections go through these cases.

Hierarchical gates definitions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The ``gate`` statement associates an identifier with a corresponding unitary matrix (or parameterized family)
transformation by a sequence of other (built-in or previously-defined) gates.

For example the ``gate`` block

.. code-block::

   gate h q {
      U(π/2, 0, π) q;
      gphase -π/4;
   }

defines a new gate called ``h`` and associates it to the unitary matrix of the Hadamard gate. Once we have
defined ``h``, we can use it in later ``gate`` blocks.

The definition does not imply that ``h`` is
implemented by an instruction ``U(π/2, 0, π)`` on the quantum computer. The implementation is up to
the user and/or compiler, given information about the instructions supported by a particular target.
A minimal compiler implementation might simply expand ``gate`` definitions repeatedly until reaching
definitions for which :ref:`defcal blocks <pulse-gates>` are known. A more sophisticated implementation
might use the `gate` definitions of the gates with associated ``defcal`` blocks to

> to build a gate library, and use methods based on KAK decompositions to rewrite into this hardware library.

Reviewer 2: nit: superfluous `to`
```suggestion
build a gate library, and use methods based on KAK decompositions to rewrite into this hardware library.
```

.. code-block::

   gate name(params) qargs
   {
     body
   }

where the optional parameter list ``params`` is a comma-separated list of variable
parameters, and the argument list ``qargs`` is a comma-separated list of qubit
arguments. The parameters are identifiers that behave as ``angle`` type with unknown
size. A compiler might recognize certain constructs and replace them with mathematically-
equivalent versions that would be true for arbitrary precision, or it might do calculations

> at a fixed angle size, for example corresponding to the size of angle parameters to the corresponding

Reviewer: ```suggestion
at a fixed ``angle`` size, for example corresponding to the size of angle parameters to the corresponding
```

.. code-block::

   gate name(params) qargs
   {
     body
   }

where the optional parameter list ``params`` is a comma-separated list of variable
parameters, and the argument list ``qargs`` is a comma-separated list of qubit
arguments. The parameters are identifiers that behave as ``angle`` type with unknown
size. A compiler might recognize certain constructs and replace them with mathematically-
equivalent versions that would be true for arbitrary precision, or it might do calculations

> at a fixed ``angle`` size, for example corresponding to the size of ``angle`` parameters to the corresponding

Reviewer: ```suggestion
at a fixed ``angle`` size, for example corresponding to the size of ``angle`` parameters in the corresponding
```

An empty body corresponds to the identity gate.

To avoid infinite recursion, gates must be declared before use and
cannot call themselves. The statement ``name(params) qargs;`` applies the gate,
and the variable parameters ``params`` can have any type that can promote to ``angle`` type.

Quantum gate modifiers
~~~~~~~~~~~~~~~~~~~~~~

A gate modifier is a keyword that applies to a gate. A modifier
:math:`m` transforms a gate :math:`U` to a new gate :math:`m(U)` acting
on the same or larger Hilbert space. We include modifiers in OpenQASM
both for programming convenience and compiler analysis.

Control modifiers
+++++++++++++++++

The modifier ``ctrl @`` replaces its gate argument :math:`U` by a
controlled-:math:`U` gate. If the control bit is 0, nothing happens to the target bit.
If the control bit is 1, :math:`U` acts on the target bit. Mathematically, the controlled-:math:`U`
gate is defined as :math:`C_U = I \otimes U^c`, where :math:`c` is the integer value of the control
bit and :math:`C_U` is the controlled-:math:`U` gate. The new quantum argument is prepended to the
argument list for the controlled-:math:`U` gate. The quantum argument can be a register, and in this

> case controlled gate broadcast over it (as for all gates //FIXME). The modified

Reviewer: What is the plan here?

Author: The plan is that all gates broadcast with the same semantics

specifies any element of :math:`U(2)` up to a
global phase [#uphase]_ . For example ``U(π/2, 0, π) q[0];``, applies a Hadamard gate to qubit ``q[0]``
(up to a non-standard global phase).

Global phase gate ``gphase``
++++++++++++++++++++++++++++

From a physical perspective, the unitaries :math:`e^{i\gamma}V` and :math:`V` are equivalent although they differ by a global
phase :math:`e^{i\gamma}`. When we add a control to these gates, however, the global phase becomes a relative phase
that is applied when the control qubit is one. A built-in global phase gate
allows the inclusion of arbitrary global phases on circuits. The instruction ``gphase(γ);`` accumulates a global phase
of :math:`e^{i\gamma}`. For example

.. code-block::

   gate X q {
      U(π, 0, π) q;
      gphase -π/2;
   }

   gate CX c, t {
      ctrl @ X c, t;
   }

defines ``CX`` as the standard CNOT gate.

Relation of the built-in gates to hardware-native gates
----------------------------

For *non-parameterized gates*, the choice of ``U`` and ``gphase`` as the built-in gates, along with one
two-qubit entangling gate CNOT as defined gives a universal gate set that can represent general n-qubit

> unitary with an :math:`O(2^n)` size description :cite:`barenco95`. This basis is not an enforced compilation

Reviewer: ```suggestion
unitaries with an :math:`O(2^n)` size description :cite:`barenco95`. This basis is not an enforced compilation
```

number of one- and two-qubit gates, giving a more compact representation
than requiring the programmer to express the full :math:`2^n \times 2^n`
matrix. However, a general :math:`n`-qubit gate can be defined using an
exponential number of these gates. Thus there is no particular privilege incurred by hardware implementations
that natively support the built-in gates.

For *parameterized gates*, the choice of built-in gates *does* constrain which hardware-native gates are well-
supported, because conversion between parameterized basis sets in general can be involved, requiring careful
selection of branch cuts and other logic that would not likely be feasible to specify as compact mathematical
expressions, nor to evaluate at runtime for cases where the parameters depend on quantum measurements.

For many current platforms the qubits are defined relative to a
rotating frame and the rotating wave approximation (RWA) holds. This is the domain covered by the OpenPulse
specification. For this case, the only supported form of run-time parameterization
will likely be via a ``rz(ϕ)`` implemented by specialized frame-tracking hardware.

> This gate is covered by the built-in ``U`` as a special case ``U(0, ϕ, θ)``

Reviewer: Do you mean to have zeros for the first two arguments of `U`?

Author: I didn't but agree that would be clearer

> This gate is covered by the built-in ``U`` as a special case ``U(0, ϕ, θ)``

Author: ```suggestion
This gate is covered by the built-in ``U`` as a special case ``U(0, 0, ϕ)``
```
